File: backend/src/main/kotlin/com/trackwatch/backend/clients/spotify/SpotifyArtistApiClient.kt
package com.trackwatch.backend.clients.spotify

import com.trackwatch.backend.exception.InternalServerErrorException
import com.trackwatch.backend.model.Artist
import com.trackwatch.backend.model.Track
import com.trackwatch.backend.model.TrackImage
import com.trackwatch.backend.utils.service.MetricService
import com.trackwatch.backend.utils.values.ErrorCode
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component
import org.springframework.web.reactive.function.client.WebClientResponseException
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Date
import java.util.Locale

@Component
class SpotifyArtistApiClient(metricService: MetricService): SpotifyApiClient(metricService) {

    private val log = LoggerFactory.getLogger(SpotifyArtistApiClient::class.java)

    fun getArtistInfo(artistId: String, accessToken: String): Artist {
        try {
            sendMetricApiCall("getArtistInfo")

            val response = webClient.get()
                .uri("/artists/$artistId")
                .header("Authorization", "Bearer $accessToken")
                .retrieve()
                .bodyToMono(Map::class.java)
                .block() ?: throw RuntimeException("No data returned from Spotify API")

            return parseArtistInfo(response)

        } catch (e: WebClientResponseException) {
            throw RuntimeException("Error while calling Spotify API: ${e.message}")

        } catch (e: Exception) {
            throw InternalServerErrorException(ErrorCode.UNHANDLED_EXCEPTION, "Error while calling Spotify API", e.toString())
        }
    }

    fun getArtistNewTracksWithRetries(artist: Artist, accessToken: String, daysLimit: Int?, page: Int, maxAttempts: Int = 3): List<Track> {
        var lastException: Exception? = null

        for (attempt in 1..maxAttempts) {
            try {
                return getArtistNewTracks(artist, accessToken, daysLimit, page)

            } catch (e: Exception) {
                lastException = e
                log.warn("Spotify API call failed (attempt $attempt/$maxAttempts): $e")

                if (attempt < maxAttempts) {
                    val waitTimeMs = 1000L * attempt // Exponential backoff
                    log.info("Retrying in ${waitTimeMs}ms")

                    try {
                        Thread.sleep(waitTimeMs)
                    } catch (ie: InterruptedException) {
                        Thread.currentThread().interrupt()
                        throw ie
                    }
                }
            }
        }

        if (lastException is WebClientResponseException) {
            throw RuntimeException("Error while calling Spotify API after $maxAttempts attempts: ${lastException.message}")
        }

        throw InternalServerErrorException(
            ErrorCode.UNHANDLED_EXCEPTION,
            "Error while calling Spotify API after $maxAttempts attempts",
            lastException?.toString() ?: ""
        )
    }

    private fun getArtistNewTracks(artist: Artist, accessToken: String, daysLimit: Int?, page: Int): List<Track> {
        val queryParams = buildQueryParams(artist.name, daysLimit, page)

        try {
            sendMetricApiCall("getArtistNewTracks")

            val response = webClient.get()
                .uri("/search?$queryParams")
                .header("Authorization", "Bearer $accessToken")
                .retrieve()
                .bodyToMono(Map::class.java)
                .block() ?: return emptyList()

            return parseTracks(response)

        } catch (e: WebClientResponseException) {
            throw RuntimeException("Error while calling Spotify API: ${e.message}")

        } catch (e: Exception) {
            throw InternalServerErrorException(ErrorCode.UNHANDLED_EXCEPTION, "Error while calling Spotify API", e.toString())
        }
    }

    private fun parseTracks(response: Map<*, *>): List<Track> {
        val tracks = response["tracks"] as Map<*, *>
        val items = tracks["items"] as List<*>

        val parsedTracks = mutableListOf<Track>()

        items.forEach { item ->
            val mapTrack = item as Map<*, *>

            val artists = mapTrack["artists"] as List<*>
            val album = mapTrack["album"] as Map<*, *>

            val releaseDateString = album["release_date"] as String
            val releaseDatePrecision = album["release_date_precision"] as String
            val releaseDate = formatReleaseDate(releaseDateString, releaseDatePrecision)

            val track = Track(
                id = mapTrack["id"] as String,
                uri = mapTrack["uri"] as String,
                name = mapTrack["name"] as String,
                artists = parseArtists(artists),
                releaseDate = releaseDate,
                isExplicit = mapTrack["explicit"] as Boolean,
                albumName= album["name"] as String,
                albumImages = parseAlbumImages(album),
                albumType = album["album_type"] as String,
                discNumber = mapTrack["disc_number"] as Int,
                albumOrder = mapTrack["track_number"] as Int,
                durationMs = mapTrack["duration_ms"] as Int
            )

            parsedTracks.add(track)
        }

        return parsedTracks
    }

    private fun formatReleaseDate(releaseDateString: String, releaseDatePrecision: String): Date {
        val dateFormat = SimpleDateFormat("yyyy-MM-dd", Locale.getDefault())

        val releaseDate = when (releaseDatePrecision) {
            "year" -> {
                val year = releaseDateString.substring(0, 4)
                dateFormat.parse("$year-01-01")
            }
            "month" -> {
                val year = releaseDateString.substring(0, 4)
                val month = releaseDateString.substring(5, 7)
                dateFormat.parse("$year-$month-01")
            }
            "day" -> {
                dateFormat.parse(releaseDateString)
            }
            else -> throw RuntimeException("Invalid release date precision: $releaseDatePrecision")
        }

        return releaseDate
    }

    private fun parseArtists(dataArtists: List<*>): List<Artist> {
        val parsedArtists = mutableListOf<Artist>()

        dataArtists.forEach { dataArtist ->
            val mapArtist = dataArtist as Map<*, *>

            val artist = Artist(
                id = mapArtist["id"] as String,
                name = mapArtist["name"] as String
            )

            parsedArtists.add(artist)
        }

        return parsedArtists
    }

    private fun parseAlbumImages(dataAlbum: Map<*, *>): List<TrackImage> {
        val parsedAlbumImages = mutableListOf<TrackImage>()

        val images = dataAlbum["images"] as List<*>

        images.forEach { dataImage ->
            val mapImage = dataImage as Map<*, *>

            val image = TrackImage(
                url = mapImage["url"] as String,
                width = mapImage["width"] as Int,
                height = mapImage["height"] as Int
            )

            parsedAlbumImages.add(image)
        }

        return parsedAlbumImages
    }

    private fun buildQueryParams(artistName: String, daysLimit: Int?, page: Int): String {
        val q = buildQuery(artistName, daysLimit)
        val type = "track"
        val limit = 50
        val offset = page * 50

        return "q=$q&type=$type&limit=$limit&offset=$offset"
    }

    private fun buildQuery(artistName: String, daysLimit: Int?): String {

        if (daysLimit == null) {
            return "artist:${artistName}"
        }

        val dateFormat = SimpleDateFormat("yyyy-MM-dd", Locale.getDefault())
        val today = Date()
        val todayIso = dateFormat.format(today)

        val calendar = Calendar.getInstance()
        calendar.time = today
        calendar.add(Calendar.DAY_OF_YEAR, - daysLimit)

        val startDate = calendar.time
        val startDateIso = dateFormat.format(startDate)

        var searchQuery = "artist:${artistName}"

        // if start_date.year == today.year
        searchQuery += if (startDateIso.substring(0, 4) == todayIso.substring(0, 4)) {
            " year:${startDateIso.substring(0, 4)}"
        } else {
            " year:${startDateIso.substring(0, 4)}-${todayIso.substring(0, 4)}"
        }

        return searchQuery
    }

    fun parseArtistInfo(response: Map<*, *>): Artist {
        val images = response["images"] as List<*>
        val imageUrl = (images.firstOrNull() as? Map<*, *>)?.get("url") as? String ?: ""

        return Artist(
            id = response["id"] as String,
            name = response["name"] as String,
            imageUrl = imageUrl
        )
    }

}

File: backend_py/trackwatch/app/clients/spotify/spotify_artist_api_client.py
from .spotify_api_client import spotify_api_request
from app.exceptions import InternalServerErrorException, ErrorCode
from app.classes import Artist, Track, TrackImage
import datetime

def get_artist_info(artist_id, access_token):
  try:
    response = spotify_api_request(
      method="GET",
      endpoint=f"/artists/{artist_id}",
      token=access_token
    )
    return parse_artist_info(response)
  except Exception as e:
    raise InternalServerErrorException(
      ErrorCode.UNHANDLED_EXCEPTION,
      f"Error while calling Spotify API: {str(e)}"
    )

def search_artist_tracks_with_retries(artist, access_token, days_limit, page, max_attempts=3):
  last_exception = None
  for attempt in range(1, max_attempts + 1):
    try:
      return search_artist_tracks(artist, access_token, days_limit, page)
    except Exception as e:
      last_exception = e
      print(f"Spotify API call failed (attempt {attempt}/{max_attempts}): {e}")
      if attempt < max_attempts:
        wait_time = 1 * attempt
        print(f"Retrying in {wait_time}s")
        import time; time.sleep(wait_time)
  raise InternalServerErrorException(
    ErrorCode.UNHANDLED_EXCEPTION,
    f"Error while calling Spotify API after {max_attempts} attempts: {str(last_exception)}"
  )

def search_artist_tracks(artist, access_token, days_limit, page):
  query_params = build_query_params(artist.name, days_limit, page)
  try:
    response = spotify_api_request(
      method="GET",
      endpoint=f"/search?{query_params}",
      token=access_token
    )
    return parse_tracks(response)
  except Exception as e:
    raise InternalServerErrorException(
      ErrorCode.UNHANDLED_EXCEPTION,
      f"Error while calling Spotify API: {str(e)}"
    )

def parse_tracks(response):
  tracks_map = response.get("tracks", {})
  items = tracks_map.get("items", [])
  parsed_tracks = []
  for map_track in items:
    artists_json = map_track.get("artists", [])
    album_json = map_track.get("album", {})
    release_date_string = album_json.get("release_date")
    release_date_precision = album_json.get("release_date_precision")
    release_date = format_release_date(release_date_string, release_date_precision)

    track = Track(
      id = map_track.get("id"),
      uri = map_track.get("uri"),
      name = map_track.get("name"),
      artists = parse_artists(artists_json),
      release_date = release_date,
      is_explicit = map_track.get("explicit"),
      album_name = album_json.get("name"),
      album_images = parse_album_images(album_json),
      album_type = album_json.get("album_type"),
      disc_number = map_track.get("disc_number"),
      album_order = map_track.get("track_number"),
      duration_ms = map_track.get("duration_ms")
    )
    parsed_tracks.append(track)
  return parsed_tracks

def format_release_date(release_date_string, release_date_precision):
  if not release_date_string or not release_date_precision:
    raise InternalServerErrorException(
      ErrorCode.UNHANDLED_EXCEPTION,
      "Invalid release date or precision"
    )
  try:
    if release_date_precision == "year":
      return datetime.datetime.strptime(f"{release_date_string[:4]}-01-01", "%Y-%m-%d")
    elif release_date_precision == "month":
      return datetime.datetime.strptime(f"{release_date_string[:7]}-01", "%Y-%m-%d")
    elif release_date_precision == "day":
      return datetime.datetime.strptime(release_date_string, "%Y-%m-%d")
    else:
      raise InternalServerErrorException(
        ErrorCode.UNHANDLED_EXCEPTION,
        f"Invalid release date precision: {release_date_precision}"
      )
  except Exception as e:
    raise InternalServerErrorException(
      ErrorCode.UNHANDLED_EXCEPTION,
      f"Invalid date parsing: {release_date_string}, {release_date_precision} -- {e}"
    )

def parse_artists(data_artists):
  parsed = []
  for map_artist in data_artists:
    artist = Artist(
      id = map_artist.get("id"),
      name = map_artist.get("name")
    )
    parsed.append(artist)
  return parsed

def parse_album_images(data_album):
  parsed = []
  images = data_album.get("images", [])
  for map_image in images:
    image = TrackImage(
      url = map_image.get("url"),
      width = map_image.get("width"),
      height = map_image.get("height")
    )
    parsed.append(image)
  return parsed

def build_query_params(artist_name, days_limit, page):
  from urllib.parse import quote_plus
  q = build_query(artist_name, days_limit)
  q_escaped = quote_plus(q)
  type_ = "track"
  limit = 50
  offset = page * 50
  return f"q={q_escaped}&type={type_}&limit={limit}&offset={offset}"

def build_query(artist_name, days_limit):
  if days_limit is None:
    return f"artist:{artist_name}"
  today = datetime.datetime.now(datetime.timezone.utc)
  start_date = today - datetime.timedelta(days=days_limit)
  start_year = start_date.strftime("%Y")
  today_year = today.strftime("%Y")
  search_query = f"artist:{artist_name}"
  if start_year == today_year:
    search_query += f" year:{start_year}"
  else:
    search_query += f" year:{start_year}-{today_year}"
  return search_query

def parse_artist_info(response):
  images = response.get("images", [])
  image_url = images[0]["url"] if images else ""
  return Artist(
    id = response.get("id"),
    name = response.get("name"),
    image_url = image_url
  )
